python/ray/serve/_private/application_state.py

```diff
@@ -109,8 +147,21 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
                 "path in your application to avoid this issue."
             )
 
+        # Delete old deployments
+        self.deployments_to_delete = set(self.deployments).difference(
+            {params["name"] for params in deployment_params}
+        )
+        for deployment_name in self.deployments_to_delete:
+            self.delete_deployment(deployment_name)
+        self.deployment_params = deployment_params
+
+        # Deploy new deployments
+        updating = []
+        for deployment in deployment_params:
+            updating.append(self.deploy_deployment(**deployment))
```

Comment:
I think for correctness this logic should happen in `update`. As written this could lead to an inconsistent internal state (e.g., if `deploy_deployment` failed, some deployments would be deleted and some would not be updated).